File: src/python/oop/ShippingLocker/locker/locker_system.py
from .locker import *
from .locker_site import LockerSite
from .notification_system import NotificationSystem
from .package import *
from .price_calculator import PriceCalculator

class LockerSystem:
    def __init__(self, location: str, locker_sizes: list[(LockerSize, int)]):
        self._location = location
        self._locker_site = LockerSite(locker_sizes)
        self._notification = NotificationSystem()
        self._price_calculator = PriceCalculator()

    def add_package(self, package: Package) -> Locker:
        locker = self._locker_site.assign_locker(package)
        if not locker:
            print("The suitable locker not found")
            return

        package.status = PackageStatus.IN_LOCKER

        self._notification.package_placed_in_locker(self._location, locker, package)

        return locker

    def get_package(self, access_code: str) -> bool:
        locker = self._locker_site.get_locker(access_code)
        if not locker:
            print("The access code is not valid")
            return False

        package = locker.package
        price = self._price_calculator.calculate_price(locker)
        package.user_account.charge_price(price)
        package.status = PackageStatus.RETRIEVED

        self._notification.package_retrieved(package)

        self._locker_site.release_locker(locker)

        return True

    def release_expired_lockers(self) -> None:
        lockers = self._locker_site.get_expired_lockers()

        for locker in lockers:
            package = locker.package
            price = self._price_calculator.calculate_price(locker)
            package.user_account.charge_price(price)
            package.status = PackageStatus.EXPIRED

            self._notification.package_expired(package)

            self._locker_site.release_locker(locker)

File: src/go/oop/ShippingLocker/internal/locker_system.go
package internal

import "fmt"

type LockerSystem struct {
	location     string
	notification *Notification
	site         *LockerManager
	calculator   *PriceCalculator
}

func System(location string, lockerConfig map[LockerSize]int) LockerSystem {
	site := Site(lockerConfig)
	return LockerSystem{
		location:     location,
		notification: &Notification{},
		site:         &site,
		calculator:   &PriceCalculator{},
	}
}

func (s *LockerSystem) AddPackage(pkg *Package) (*Locker, error) {
	locker, err := s.site.AssignLocker(pkg)
	if err != nil {
		return nil, err
	}

	pkg.Status = InLocker
	s.notification.PackagePlaced(s.location, locker, *pkg)

	return locker, nil
}

func (s *LockerSystem) GetPackage(code string) (*Package, error) {
	locker := s.site.GetLocker(code)
	if locker == nil {
		return nil, fmt.Errorf("access code not valid")
	}

	price := s.calculator.Calculate(locker)

	pkg, err := s.site.ReleaseLocker(locker)
	if err != nil {
		return nil, err
	}

	pkg.User.BillingAmount += price
	pkg.Status = Delivered

	s.notification.PackageRetrieved(*pkg)

	return pkg, nil
}

func (s *LockerSystem) ReleaseExpiredPackages() {
	lockers := s.site.GetExpiredLockers()

	for _, locker := range lockers {
		price := s.calculator.CalculateExpirationPrice(locker)

		pkg, err := s.site.ReleaseLocker(locker)
		if err != nil {
			fmt.Println(err)
			continue
		}

		pkg.User.BillingAmount += price
		pkg.Status = Expired

		s.notification.PackageExpired(*pkg)
	}
}
